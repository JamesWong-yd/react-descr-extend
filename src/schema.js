import docPropType from './docPropType';

export default function schema(component, metadata = {}) {
  if (!component) {
    throw new Error('schema: component is required');
  }
  if (!metadata.description) {
    throw new Error('schema: description is required');
  }

  const reactDesc = { ...metadata };
  delete reactDesc.props;
  let defaultProps;
  let propTypes;

  if (metadata.props) {
    Object.keys(metadata.props).forEach((propName) => {
      const prop = metadata.props[propName];
      // if it is an array it means that it is a react schema instance
      if (Array.isArray(prop)) {
        const validate = prop[0];
        const description = prop[1];
        const options = prop[2];

        if (options && options.defaultProp) {
          if (!defaultProps) {
            defaultProps = {};
          }
          defaultProps[propName] = options.defaultProp;
          delete options.defaultProp;
        }

        if (!propTypes) {
          propTypes = {};
        }
        propTypes[propName] = docPropType(validate, description, options);
      } else {
        propTypes[propName] = prop;
      }
    });
  }

  /* eslint-disable no-param-reassign */
  if (process.env.NODE_ENV !== 'production') {
    // only add documentation if it is not production
    component.$$reactDesc = reactDesc;
    if (propTypes) {
      component.propTypes = propTypes;
    }
  }

  // but always add default props :)
  if (!component.defaultProps && defaultProps) {
    component.defaultProps = defaultProps;
  }
  /* eslint-enable no-param-reassign */
}
