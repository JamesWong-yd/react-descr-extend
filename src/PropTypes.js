export default {
  any: { type: 'any' },
  array: { type: 'array' },
  bool: { type: 'bool' },
  element: { type: 'element' },
  func: { type: 'func' },
  node: { type: 'node' },
  number: { type: 'number' },
  object: { type: 'object' },
  string: { type: 'string' },
  symbol: { type: 'symbol' },
  arrayOf: args => ({ type: 'arrayOf', args }),
  instanceOf: args => ({ type: 'instanceOf', args }),
  objectOf: args => ({ type: 'objectOf', args }),
  oneOfType: args => ({ type: 'oneOfType', args }),
  oneOf: args => ({ type: 'oneOf', args }),
  shape: args => ({ type: 'shape', args }),
};
