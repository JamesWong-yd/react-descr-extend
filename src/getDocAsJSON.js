function buildShapeFormat(validate) {
  const format = {};
  Object.keys(validate.args).forEach((arg) => {
    const element = validate.args[arg];
    if (element.type === 'shape') {
      format[arg] = buildShapeFormat(element, true);
    } else {
      format[arg] = getFormat(element);
    }
  });
  return format;
}

function getFormat(validate) {
  switch (validate.type) {
    case 'arrayOf':
      return `[${validate.args.type}, ...]`;
    case 'bool':
      return 'boolean';
    case 'func':
      return 'function';
    case 'instanceOf':
      return `new ${validate.args.name}(...)`;
    case 'objectOf':
      return `{ test: ${validate.args.type}, ... }`;
    case 'oneOf':
      return validate.args;
    case 'oneOfType':
      return validate.args.map(
        arg => getFormat(arg)
      );
    case 'shape':
      return JSON.stringify(buildShapeFormat(validate));
    default:
      return validate.type ? validate.type : 'custom';
  }
}

function getPropTypeAsJson(propType, propName, defaultValue) {
  const {
    deprecated, format, required, ...extras
  } = propType.$$reactDesc.options;
  const { validate } = propType.$$reactDesc;
  const documentation = {
    description: propType.$$reactDesc.description,
    name: propName,
  };

  if (defaultValue) {
    documentation.defaultValue = defaultValue;
  }

  if (deprecated) {
    documentation.deprecated = deprecated;
  }

  if (required) {
    documentation.required = required;
  }

  if (Object.keys(extras).length > 0) {
    documentation.extras = extras;
  }

  documentation.format = format || getFormat(validate);

  return documentation;
}

export default function getDocAsJSON(component) {
  if (!component) {
    throw new Error('getDocAsJSON: component is required');
  }

  const documentation = {
    name: component.displayName || component.name,
  };
  if (component.$$reactDesc) {
    Object.assign(documentation, component.$$reactDesc);

    if (component.propTypes) {
      const propTypes = [];
      Object.keys(component.propTypes).forEach((propName) => {
        const propType = component.propTypes[propName];
        if (propType.$$reactDesc) {
          propTypes.push(
            getPropTypeAsJson(
              propType, propName, component.defaultProps[propName]
            )
          );
        }
      });
      if (propTypes.length > 0) {
        documentation.properties = propTypes;
      }
    }
  }
  return documentation;
}
