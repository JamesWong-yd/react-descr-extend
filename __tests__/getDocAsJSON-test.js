import React from 'react';
import ReactPropTypes from 'prop-types';
import getDocAsJSON from '../src/getDocAsJSON';
import schema from '../src/schema';
import PropTypes from '../src/PropTypes';

class FakeComponent {}
const complexShape = {
  color: PropTypes.string,
  fontSize: PropTypes.number,
  one: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.instanceOf(FakeComponent),
    (props, propName) => typeof props[propName] === 'number',
  ]),
  test: PropTypes.instanceOf(FakeComponent),
  test2: PropTypes.shape({
    test3: PropTypes.string,
    test4: PropTypes.shape({
      test5: PropTypes.string,
      test6: PropTypes.number,
      test7: (props, propName) => typeof props[propName] === 'number',
      test8: PropTypes.arrayOf(PropTypes.string),
    }),
  }),
};

const DocumentedComponent = () => <div>fake</div>;

const DeprecatedComponent = () => <div>deprecated</div>;
DeprecatedComponent.propTypes = {
  testDeprecated: ReactPropTypes.string,
};
const NoPropTypeComponent = () => <div>fake</div>;
/* eslint-disable react/prefer-es6-class, react/prefer-stateless-function */
const ExtraInfoComponent = React.createClass({
  render() {
    return (
      <div>fake</div>
    );
  },
});
/* eslint-enable react/prefer-es6-class, react/prefer-stateless-function */

schema(DocumentedComponent, {
  description: 'component',
  props: {
    test: [PropTypes.any, 'any'],
    test2: [PropTypes.array, 'array'],
    test3: [
      PropTypes.arrayOf(PropTypes.number), 'arrayOf',
    ],
    test4: [PropTypes.bool, 'bool'],
    test5: [() => {}, 'custom', {
      format: 'XXX-XX-XXXX',
    }],
    test6: [PropTypes.element, 'element'],
    test7: [PropTypes.func, 'func'],
    test8: [PropTypes.instanceOf(FakeComponent), 'instanceOf'],
    test9: [PropTypes.node, 'node'],
    test10: [PropTypes.number, 'number'],
    test11: [PropTypes.object, 'object'],
    test12: [PropTypes.objectOf(PropTypes.number), 'objectOf'],
    test13: [PropTypes.oneOf(['News', 'Photos']), 'oneOf'],
    test14: [PropTypes.oneOfType([
      PropTypes.number, PropTypes.string,
    ]), 'oneOfType'],
    test15: [PropTypes.symbol, 'symbol'],
    test16: [PropTypes.shape(complexShape), 'shape'],
    testRequired: [PropTypes.string, 'testRequired', {
      required: true,
    }],
    testDeprecated: [PropTypes.string, 'testDeprecated', {
      defaultProp: 'abc',
      deprecated: 'use something else',
    }],
    testExtra: [PropTypes.string, 'testExtra', {
      defaultProp: 'abc',
      usage: 'test5 = "abc"',
    }],
    testNative: ReactPropTypes.string,
  },
});
schema(NoPropTypeComponent, {
  description: 'noPropType',
});
schema(DeprecatedComponent, {
  description: 'component',
  deprecated: 'use button instead',
});
schema(ExtraInfoComponent, {
  description: 'component',
  props: {
    test: [PropTypes.string, 'test', {
      defaultProp: 'abc',
    }],
  },
  usage: 'import component  from "grommet/components/Anchor"',
});

it('fails for missing component property', () => {
  expect(() => {
    getDocAsJSON(undefined);
  }).toThrowError('getDocAsJSON: component is required');
});

it('documents empty doc for component without reactDesc', () => {
  const documentation = getDocAsJSON(FakeComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents empty propType doc for component', () => {
  const documentation = getDocAsJSON(NoPropTypeComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents a basic documented component', () => {
  const documentation = getDocAsJSON(DocumentedComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents a deprecated documented component', () => {
  const documentation = getDocAsJSON(DeprecatedComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents an extra info documented component', () => {
  const documentation = getDocAsJSON(ExtraInfoComponent);
  expect(documentation).toMatchSnapshot();
});
