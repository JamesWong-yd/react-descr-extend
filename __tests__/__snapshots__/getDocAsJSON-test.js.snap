exports[`test documents a basic documented component 1`] = `
Object {
  "description": "component",
  "name": "DocumentedComponent",
  "properties": Array [
    Object {
      "description": "any",
      "format": "any",
      "name": "test",
    },
    Object {
      "description": "array",
      "format": "array",
      "name": "test2",
    },
    Object {
      "description": "arrayOf",
      "format": "[number, ...]",
      "name": "test3",
    },
    Object {
      "description": "bool",
      "format": "boolean",
      "name": "test4",
    },
    Object {
      "description": "custom",
      "format": "XXX-XX-XXXX",
      "name": "test5",
    },
    Object {
      "description": "element",
      "format": "element",
      "name": "test6",
    },
    Object {
      "description": "func",
      "format": "function",
      "name": "test7",
    },
    Object {
      "description": "instanceOf",
      "format": "new FakeComponent(...)",
      "name": "test8",
    },
    Object {
      "description": "node",
      "format": "node",
      "name": "test9",
    },
    Object {
      "description": "number",
      "format": "number",
      "name": "test10",
    },
    Object {
      "description": "object",
      "format": "object",
      "name": "test11",
    },
    Object {
      "description": "objectOf",
      "format": "{ test: number, ... }",
      "name": "test12",
    },
    Object {
      "description": "oneOf",
      "format": "News&#124;Photos",
      "name": "test13",
    },
    Object {
      "description": "oneOfType",
      "format": "number&#124;string",
      "name": "test14",
    },
    Object {
      "description": "symbol",
      "format": "symbol",
      "name": "test15",
    },
    Object {
      "description": "shape",
      "format": "{\"color\":\"string\",\"fontSize\":\"number\",\"one\":\"string&#124;number&#124;new FakeComponent(...)&#124;custom\",\"test\":\"new FakeComponent(...)\",\"test2\":{\"test3\":\"string\",\"test4\":{\"test5\":\"string\",\"test6\":\"number\",\"test7\":\"custom\",\"test8\":\"[string, ...]\"}}}",
      "name": "test16",
    },
    Object {
      "description": "testRequired",
      "format": "string",
      "name": "testRequired",
      "required": true,
    },
    Object {
      "deprecated": "use something else",
      "description": "testDeprecated",
      "format": "string",
      "name": "testDeprecated",
    },
    Object {
      "defaultValue": "abc",
      "description": "testExtra",
      "extras": Object {
        "usage": "test5 = \"abc\"",
      },
      "format": "string",
      "name": "testExtra",
    },
  ],
}
`;

exports[`test documents a deprecated documented component 1`] = `
Object {
  "deprecated": "use button instead",
  "description": "component",
  "name": "DeprecatedComponent",
}
`;

exports[`test documents an extra info documented component 1`] = `
Object {
  "description": "component",
  "extras": Object {
    "usage": "import component  from \"grommet/components/Anchor\"",
  },
  "name": "ExtraInfoComponent",
  "properties": Array [
    Object {
      "defaultValue": "abc",
      "description": "test",
      "format": "string",
      "name": "test",
    },
  ],
}
`;

exports[`test documents empty doc for component without reactDesc 1`] = `
Object {
  "name": "FakeComponent",
}
`;

exports[`test documents empty propType doc for component 1`] = `
Object {
  "description": "noPropType",
  "name": "NoPropTypeComponent",
}
`;
